src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs

```diff
@@ -13,115 +13,236 @@ namespace NuGet.Packaging.Signing
+                }
+                else
                 {
-                    if (settings.AlwaysVerifyCountersignature || ShouldFallbackToRepositoryCountersignature(primarySummary))
+                    verifySettings = new SignatureVerifySettings(
+                        allowIllegal: settings.AllowIllegal,
+                        allowUntrusted: settings.AllowUntrusted,
+                        reportUntrusted: true,
+                        allowUnknownRevocation: settings.AllowUnknownRevocation,
+                        reportUnknownRevocation: settings.ReportUnknownRevocation);
+
+                    var countersignatureSummary = VerifyValidityAndTrust(countersignature, settings, verifySettings, certificateExtraStore, issues);
+
+                    if (primarySummary == null)
+                    {
+                        status = countersignatureSummary.Status;
+                    }
+                    else if (PrimarySignatureNeedsTimestamp(primarySummary))
                     {
-                        var countersignature = RepositoryCountersignature.GetRepositoryCountersignature(signature);
-                        verifySettings = new SignatureVerifySettings(
-                            treatIssuesAsErrors: !settings.AllowIllegal,
-                            allowUntrustedRoot: false,
-                            allowUnknownRevocation: settings.AllowUnknownRevocation,
-                            logOnSignatureExpired: true);
+                        if (countersignatureSummary.Status == SignatureVerificationStatus.Valid &&
+                            countersignatureSummary.Timestamp != null &&
+                            Rfc3161TimestampVerificationUtility.ValidateSignerCertificateAgainstTimestamp(signature.SignerInfo.Certificate, countersignatureSummary.Timestamp))
+                        {
+                            var index = issues.FindIndex(log => log.Code == NuGetLogCode.NU3037);
 
-                        var counterSummary = VerifyValidityAndTrust(countersignature, settings, verifySettings, certificateExtraStore, issues);
-                        status = counterSummary.Status;
+                            if (index > -1)
+                            {
+                                issues.RemoveAt(index);
+                            }
 
-                        if (!Rfc3161TimestampVerificationUtility.ValidateSignerCertificateAgainstTimestamp(signature.SignerInfo.Certificate, counterSummary.Timestamp))
+                            status = SignatureVerificationStatus.Valid;
+                        }
+                        else
                         {
-                            issues.Add(SignatureLog.Issue(!settings.AllowIllegal, NuGetLogCode.NU3011, string.Format(CultureInfo.CurrentCulture, Strings.VerifyError_SignatureNotTimeValid, signature.FriendlyName)));
-                            status = SignatureVerificationStatus.Illegal;
+                            issues.Add(
```

Comment:
Given that, if the primary signature is expired it will log an issue, wouldn't this issue be logged twice?